### PR TITLE
MBS-13528 / MBS-13696: Calculate invalid edit notes in more places and with more invisible characters

### PR DESCRIPTION
**Pull request by reosarevok**
### Implement MBS-13528 and MBS-13696

# Problem
People are using Hangul fillers to circumvent the edit note requirement. Also, while 63898c73fe4482688bee7c57e937035ad5c76b0a wanted to check all edit notes, we forgot that some places that go through `ws/js` are not affected by it.  

# Solution
HANGUL FILLER is a character often used to have "empty usernames" and similar things since it's an empty space but not considered a space char. This considers it, together with other similar Hangul fillers, insufficient for a valid note.

While testing in the release editor I noticed something was weird since Hangul-filler only edit notes were not getting rejected. The second commit here fixes that bug (we were not checking the right variable).

The third commit makes it so that invisible characters are not actually stripped by `sanitize`, since they can have value when they are used as part of text, just not as the only content of an edit note.

The fourth commit also adds `BRAILLE PATTERN BLANK` to the list of invisible characters, since it similarly doesn't actually have any use other than being invisible when used outside a Braille context.

The last commit here also makes it so that the relationship editor and editing in the release editor also check for invalid notes.

# Testing
Tested manually both for artist and release edits (since release edits use the JS version of this). Added tests for them to the perl `Validation` tests. Also tested the last commit manually separately by entering `⠀ᅟ` as the edit note when editing a release and when in the relationship editor.

### CODE REVIEW

File: lib/MusicBrainz/Server/Data/Utils.pm [hunk 1]
```diff
@@ -447,7 +447,7 @@ sub remove_lineformatting_characters {
     # - zwsp
     # - shy
     # - Other, control (including TAB \x09, LF \x0A, and CR \x0D)
-    =~ s/[\N{ZERO WIDTH SPACE}\N{SOFT HYPHEN}\p{Cc}]//gr;
+    =~ s/[\N{ZERO WIDTH SPACE}\N{SOFT HYPHEN}\N{HANGUL FILLER}\N{HALFWIDTH HANGUL FILLER}\N{HANGUL JUNGSEONG FILLER}\N{HANGUL CHOSEONG FILLER}\p{Cc}]//gr;
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Data/Utils.pm:450]: Perhaps we should start using `/grxx` here to allow (ignored) whitespace/newlines within the character class.

**Review by mwiencek — APPROVED** [review]

File: lib/MusicBrainz/Server/Data/Utils.pm (hunk 1 above)
Comment (reosarevok) [reply lib/MusicBrainz/Server/Data/Utils.pm:450]: I also did consider having a separate line for the Hangul bit, maybe that'd be better.

Comment (yvanzo) [reply lib/MusicBrainz/Server/Data/Utils.pm:450]: Is that a WIP or a PR ready to merge?

Comment (reosarevok) [reply lib/MusicBrainz/Server/Data/Utils.pm:450]: It should be ready to merge, unless one of you two prefers splitting this into multiple lines or multiple regexes already now.

Comment (yvanzo) [reply lib/MusicBrainz/Server/Data/Utils.pm:450]: Actually I just spotted an issue: The code comments above have to be updated as well.

**Review by yvanzo — CHANGES_REQUESTED** [review]

File: lib/MusicBrainz/Server/Data/Utils.pm (hunk 1 above)
Comment (yvanzo) [reply lib/MusicBrainz/Server/Data/Utils.pm:450]: > Perhaps we should start using `/grxx` here to allow (ignored) whitespace/newlines within the character class.

@mwiencek: Not sure what you mean `/grxx`. A documentation link would be helpful.

Comment (mwiencek) [reply lib/MusicBrainz/Server/Data/Utils.pm:450]: Specifically the `xx` flag: https://perldoc.perl.org/perlre#/x-and-/xx (the existing regexp is already using `gr`, hence `grxx`).

Comment (reosarevok) [reply lib/MusicBrainz/Server/Data/Utils.pm]: I made some changes for clarity and future grow-ability, plus added `BRAILLE PATTERN BLANK` to the dodgy list.

**Comment by mwiencek** [comment]
This seems to work for me when testing on the release group edit form, but not in the release editor (I can still submit an edit note containing only hangul filler characters there). (I can also submit an edit note containing only "a" in the release editor, which I thought wasn't possible...)

**Comment by reosarevok** [comment]
Huh. Both of those are blocked for me locally, so I'm not sure what's going on?

**Review by yvanzo — APPROVED** [review]
Successfully tested with the braille blank pattern and the four Hangul fillers separately, in both release editor ad release group editor.

For tests I used https://invisible-characters.com/ to copy/paste those characters. It lists some more invisible characters that we might be handling at the moment.

**Comment by mwiencek** [comment]
> Huh. Both of those are blocked for me locally, so I'm not sure what's going on?

We figured this out on IRC already, but just to ensure it's clarified here: the issue was just that there was no validation performed when editing releases, which you've fixed in https://github.com/metabrainz/musicbrainz-server/pull/3315/commits/26490fbfb7fc478663f334158cfa43045cb072ef.

**Review by mwiencek — APPROVED** [review]
Looks good now, thanks for the making the related improvements. :)

**Pull request merged**